pkg/backplane/ca/ca.go

```diff
@@ -0,0 +1,178 @@
+	validFrom := time.Now()
+	tmpl := x509.Certificate{
+		SerialNumber:          serialNumber,
+		Subject:               pkixName,
+		NotBefore:             validFrom,
+		NotAfter:              validFrom.Add(validFor),
+		KeyUsage:              0,
+		ExtKeyUsage:           []x509.ExtKeyUsage{},
+		BasicConstraintsValid: true,
+	}
+	mut(&tmpl)
+	parent := a.cert
+	// NewAuthority case
+	if a.cert == nil {
+		parent = &tmpl
+	}
+	der, err := x509.CreateCertificate(randReader, &tmpl, parent, pub, a.privKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "coouldn't make authority certificate")
+	}
+	cert, err := x509.ParseCertificate(der)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't re-parse created certificate")
+	}
+	return cert, nil
+}
+
+func (a *Authority) nextSerial() *big.Int {
+	if a.lastSerial == nil {
+		num := big.NewInt(1)
+		a.lastSerial = num
+		return num
+	}
+	// Make a copy
+	return a.lastSerial.Add(a.lastSerial, big.NewInt(1))
+}
+
+// CACert is the root certificate of the CA.
+func (a *Authority) CACert() *x509.Certificate {
+	return a.cert
+}
+
+func (a *Authority) CACertPool() *x509.CertPool {
+	pool := x509.NewCertPool()
+	pool.AddCert(a.CACert())
+	return pool
+}
+
+// ServerKey makes a client cert signed by out root CA. The returned certificate
+// has a chain including the root CA cert.
+func (a *Authority) ServerKey(name string) (*tls.Certificate, error) {
+	privKey, err := rsa.GenerateKey(randReader, rsaBits)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't generate private key")
+	}
+	cert, err := a.makeCert(privKey.Public(), func(cert *x509.Certificate) {
+		cert.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
+		ip := net.ParseIP(name)
+		if ip != nil {
+			cert.IPAddresses = []net.IP{ip}
+		} else {
+			cert.DNSNames = []string{name}
+		}
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't make server certificate")
+	}
+	return &tls.Certificate{
+		Certificate: [][]byte{cert.Raw, a.cert.Raw},
+		PrivateKey:  privKey,
+		Leaf:        cert,
+	}, nil
+}
+
+// MakeServerConfig makes a new server certificate, then returns a TLS config that uses it
+// and will verify peer certificates
+func (a *Authority) MakeServerConfig(name string) (*tls.Config, error) {
+	cert, err := a.ServerKey(name)
+	if err != nil {
+		return nil, err
+	}
+
+	pool := x509.NewCertPool()
+	pool.AddCert(a.cert)
+
+	return &tls.Config{
+		Certificates: []tls.Certificate{*cert},
+		ServerName:   name,
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+		ClientCAs:    pool,
+	}, nil
+}
+
+// ClientKey makes a client cert signed by out root CA. The returned certificate
+// has a chain including the root CA
+func (a *Authority) ClientKey(name string) (*tls.Certificate, error) {
+	privKey, err := rsa.GenerateKey(randReader, rsaBits)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't generate private key")
+	}
+	cert, err := a.makeCert(privKey.Public(), func(cert *x509.Certificate) {
```

Comment:
the `mut` function seems to be the only difference in the `ServerKey` and `ClientKey` functions. Maybe there's another way to organize this so we can combine the common code and pass a different callback?

I suggest this because the wrapped errors are identical and it has the potential to be confusing if there is no stack to go with it